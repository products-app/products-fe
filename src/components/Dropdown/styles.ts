export default {
  container: 'relative inline-block text-left',
  buttonOpen:
    'inline-flex justify-center w-full rounded-full border-none hover:bg-gray800 px-3 py-3 text-lg leading-5 font-medium text-gray-700 hover:text-gray-500 focus:outline-none focus:border-blue-300 focus:shadow-outline-blue active:bg-gray600 active:text-white transition ease-in-out duration-150',
  list: 'origin-top-right absolute left-0 mt-2 w-56 rounded-md shadow-lg',
  listDisplayNone:
    'origin-top-right absolute left-0 mt-2 w-56 rounded-md shadow-lg hidden',
  link: 'block px-4 py-2 text-sm leading-5 text-gray200 hover:bg-gray600 hover:text-white focus:outline-none focus:bg-gray-100 focus:text-gray-900 flex gap-2 items-center',
}

import { styled } from '@lebernardo/react'

export const LinksContainer = styled('div', {
  background: '$gray900',
  borderRadius: '0.375rem',
})
