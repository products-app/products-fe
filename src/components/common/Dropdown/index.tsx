import { useEffect, useRef, useState } from 'react'
import classNames from 'classnames'
import { UserCircle } from 'phosphor-react'
import {
  DropdownContainer,
  ButtonDropdown,
  ListContainer,
  LinksContainer,
} from './styles'

type DropdownProps = {
  items: app.DropdownItem[]
}

const Dropdown = ({ items }: DropdownProps) => {
  const wrapperRef = useRef<HTMLDivElement>(null)
  const [display, setDisplay] = useState(false)

  const handleDisplay = () => {
    setDisplay(!display)
  }

  useEffect(() => {
    function handleClickOutside(event: Event) {
      if (
        wrapperRef.current &&
        !wrapperRef.current.contains(event.target as HTMLElement)
      ) {
        setDisplay(false)
      }
    }
    document.addEventListener('mousedown', handleClickOutside)
    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, [wrapperRef])

  return (
    <DropdownContainer ref={wrapperRef}>
      <ButtonDropdown onClick={handleDisplay}>
        <UserCircle />
      </ButtonDropdown>
      <ListContainer
        className={classNames({
          show: display,
        })}
      >
        <LinksContainer>
          {items.map((item) => (
            <a key={item.label} href={item.link}>
              {item.icon && item.icon} {item.label}
            </a>
          ))}
        </LinksContainer>
      </ListContainer>
    </DropdownContainer>
  )
}

const defaultProps = {
  items: [],
}

Dropdown.defaultProps = defaultProps
export default Dropdown
