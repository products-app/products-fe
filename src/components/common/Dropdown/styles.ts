import { styled } from '@lebernardo/react'

export const DropdownContainer = styled('div', {
  position: 'relative',
  display: 'inline-block',
  textAlign: 'left',
})

export const ButtonDropdown = styled('button', {
  display: 'inline-flex',
  justifyContent: 'center',
  width: '100%',
  borderRadius: '9999px',
  padding: 0,
  margin: 0,

  '&:hover': {
    background: '$gray800',
    color: '$gray500',
  },

  svg: {
    width: '1.15rem',
    height: '1.15rem',
    color: '$white',
    fontSize: '$md',
  },
})

export const ListContainer = styled('div', {
  position: 'absolute',
  borderRadius: '0.375rem',
  visibility: 'hidden',
})

export const LinksContainer = styled('div', {
  background: '$gray900',
  transformOrigin: 'top right',
  left: 0,
  borderRadius: '0.375rem',
  filter:
    'drop-shadow(0 10px 8px rgb(0 0 0 / 0.04)) drop-shadow(0 4px 3px rgb(0 0 0 / 0.1))',
  width: '14rem',
  overflow: 'hidden',

  div: {
    padding: '0.25rem 0',
  },

  a: {
    color: '$gray200',
    display: 'flex',
    gap: '.5rem',
    alignItems: 'center',
    padding: '.5rem 1rem',
    fontSize: '$sm',

    '&:hover': {
      color: '$white',
      background: '$base700',
    },
  },
})
