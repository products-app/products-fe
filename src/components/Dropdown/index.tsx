import classNames from 'classnames'
import { UserCircle } from 'phosphor-react'
import { useState } from 'react'
import styles from './styles'

type DropdownProps = {
  items: app.DropdownItem[]
}

const Dropdown = ({ items }: DropdownProps) => {
  const [display, setDisplay] = useState(false)

  const handleDisplay = () => {
    setDisplay(!display)
  }

  return (
    <div className={styles.container} suppressHydrationWarning>
      <div suppressHydrationWarning>
        <span className="rounded-md shadow-sm">
          <button
            type="button"
            className={styles.buttonOpen}
            onClick={handleDisplay}
          >
            <UserCircle className="text-white text-lg" />
          </button>
        </span>
      </div>
      <div
        className={classNames(styles.list, {
          [styles.listDisplayNone]: !display,
        })} 
        suppressHydrationWarning
      >
        <div className="rounded-md bg-gray900 shadow-xs">
          <div className="py-1">
            {items.map((item) => (
              <a
                key={item.label}
                href={item.link}
                className={styles.link}
                onClick={item.onClick && item.onClick}
              >
                {item.icon && item.icon} {item.label}
              </a>
            ))}
          </div>
        </div>
      </div>
    </div>
  )
}

const defaultProps = {
  items: [],
}

Dropdown.defaultProps = defaultProps
export { Dropdown }
