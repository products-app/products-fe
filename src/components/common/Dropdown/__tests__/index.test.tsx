import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Dropdown from '../index'

const dropdownItems = [
  {
    label: 'Meus pedidos',
    link: '/user/orders',
  },
  {
    label: 'Fazer logout',
    link: '/users/logout',
    icon: <div />,
  },
]
describe('Loading component', () => {
  it('should render correctly', () => {
    const { container } = render(<Dropdown />)
    expect(container).toMatchSnapshot()
  })

  it('should render dropdown with items', () => {
    render(<Dropdown items={dropdownItems} />)
    expect(screen.getByText('Meus pedidos')).toBeInTheDocument()
  })

  it('should click the button to open dropdown', async () => {
    render(<Dropdown items={dropdownItems} />)

    const handleClick = jest.fn()
    const buttonElement = screen.getByRole('button')
    buttonElement.addEventListener('click', handleClick, false)

    await userEvent.click(buttonElement)

    expect(handleClick).toHaveBeenCalled()
  })
})
