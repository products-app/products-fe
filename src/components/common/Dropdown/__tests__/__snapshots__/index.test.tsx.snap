// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading component should render correctly 1`] = `
<div>
  <div
    class="c-lXviy"
  >
    <button
      class="c-khCkZv"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <circle
          cx="128"
          cy="128"
          fill="none"
          r="96"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
        <circle
          cx="128"
          cy="120"
          fill="none"
          r="40"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
        <path
          d="M63.8,199.4a72,72,0,0,1,128.4,0"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
    </button>
    <div
      class="c-fwKBLg"
    >
      <div
        class="c-biAswx"
      />
    </div>
  </div>
</div>
`;
